google/oauth2/flow.py

```diff
@@ -256,3 +266,175 @@ class using this flow's :attr:`credentials`.
+                prompting the user for the authorization code.
+            kwargs: Additional key-word arguments passed through to
+                :meth:`authorization_url`.
+
+        Returns:
+            google.oauth2.credentials.Credentials: The OAuth 2.0 credentials
+                for the user.
+        """
+        kwargs.setdefault('prompt', 'consent')
+
+        self.redirect_uri = self._OOB_REDIRECT_URI
+
+        auth_url, _ = self.authorization_url(**kwargs)
+
+        print(authorization_prompt_message.format(url=auth_url))
+
+        code = input(authorization_code_message)
+
+        self.fetch_token(code=code)
+
+        return self.credentials
+
+    def run_local_server(
+            self, host='localhost', port=8080,
+            authorization_prompt_message=_DEFAULT_AUTH_PROMPT_MESSAGE,
+            success_message=_DEFAULT_WEB_SUCCESS_MESSAGE,
+            **kwargs):
+        """Run the flow using the server strategy.
+
+        The server strategy instructs the user to open the authorization URL in
+        their browser and will attempt to automatically open the URL for them.
+        It will start a local web server to listen for the authorization
+        response. Once authorization is complete the authorization server will
+        redirect the user's browser to the local web server. The web server
+        will get the authorization code from the response and shutdown. The
+        code is then exchanged for a token.
+
+        Args:
+            host (str): The web host for the local redirect server.
+            port (int): The port for the local redirect server.
+            authorization_prompt_message (str): The message to display to tell
+                the user to navigate to the authorization URL.
+            success_message (str): The message to display in the web browser
+                the authorization flow is complete.
+            kwargs: Additional key-word arguments passed through to
+                :meth:`authorization_url`.
+
+        Returns:
+            google.oauth2.credentials.Credentials: The OAuth 2.0 credentials
+                for the user.
+        """
+        self.redirect_uri = 'http://{}:{}/'.format(host, port)
+
+        auth_url, _ = self.authorization_url(**kwargs)
+
+        local_server = self._LocalRedirectServer(
+            success_message, (host, port), self._LocalRedirectRequestHandler)
+
+        webbrowser.open(auth_url, new=1, autoraise=True)
+
+        print(authorization_prompt_message.format(url=auth_url))
+
+        local_server.handle_request()
+
+        authorization_response = urllib.parse.urljoin(
+            # Note: using https here because oauthlib is very picky that
+            # OAuth 2.0 should only occur over https.
+            'https://localhost', local_server.last_request_path)
+        self.fetch_token(authorization_response=authorization_response)
+
+        return self.credentials
+
+    class _LocalRedirectServer(BaseHTTPServer.HTTPServer):
```

Comment:
I am BIGLY :-1: on classes defined anywhere other than top-level. Why is this defined here rather than at module scope?